readthedocs/doc_builder/python_environments.py

```diff
@@ -72,6 +74,112 @@ def delete_existing_venv_dir(self):
+                *cmd,
+            )
+
+        for tool, version in self.config.build.tools.items():
+            version = version.full_version  # e.g. 3.9 -> 3.9.7
+
+            # TODO: generate the correct path for the Python version
+            # tool_path = f'{self.config.build.os}/{tool}/2021-08-30/{version}.tar.gz'
+            tool_path = f'{self.config.build.os}-{tool}-{version}.tar.gz'
+            tool_version_cached = build_tools_storage.exists(tool_path)
+            if tool_version_cached:
+                remote_fd = build_tools_storage.open(tool_path, mode='rb')
+                with tarfile.open(fileobj=remote_fd) as tar:
+                    # Extract it on the shared path between host and Docker container
+                    extract_path = os.path.join(self.project.doc_path, 'tools')
+                    tar.extractall(extract_path)
+
+                    # Move the extracted content to the ``asdf`` installation
+                    cmd = [
+                        'mv',
+                        f'{extract_path}/{version}',
+                        os.path.join(
+                            settings.RTD_DOCKER_WORKDIR,
+                            f'.asdf/installs/{tool}/{version}',
+                        ),
+                    ]
+                    self.build_env.run(
+                        *cmd,
+                    )
+            else:
+                log.debug(
+                    'Cached version for tool not found. os=%s tool=%s version=% filename=%s',
+                    self.config.build.os,
+                    tool,
+                    version,
+                    tool_path,
+                )
+                # If the tool version selected is not available from the
+                # cache we compile it at build time
+                cmd = [
+                    # TODO: make this environment variable to work
+                    # 'PYTHON_CONFIGURE_OPTS="--enable-shared"',
+                    'asdf',
+                    'install',
+                    tool,
+                    version,
+                ]
+                self.build_env.run(
+                    *cmd,
+                )
+
+            # Make the tool version chosen by the user the default one
+            cmd = [
+                'asdf',
+                'global',
+                tool,
+                version,
+            ]
+            self.build_env.run(
+                *cmd,
+            )
+
+            # Recreate shims for this tool to make the new version
+            # installed available
+            cmd = [
+                'asdf',
+                'reshim',
+                tool,
+            ]
+            self.build_env.run(
+                *cmd,
+            )
+
+            if all([
+                    tool == 'python',
+                    # Do not install them if the tool version was cached
+                    not tool_version_cached,
+                    # Do not install them on conda/mamba
```

Comment:
This comment is a bit more useful (explains what this logic is excluding). Is there not a more explicit way to exclude conda (eg. `not self.config.conda`)?  

I see this code below:

`self.config.python_interpreter in ('conda', 'mamba'),`

We should use the same logic.